rmw_opensplice_cpp/src/rmw_take.cpp

```diff
@@ -79,6 +81,157 @@ take(
+      error_string = "_DataReader.take failed with: an internal error has occurred";
+      break;
+    case DDS::RETCODE_ALREADY_DELETED:
+      error_string = "_DataReader.take failed with: this DataReader has already been deleted";
+      break;
+    case DDS::RETCODE_OUT_OF_RESOURCES:
+      error_string = "_DataReader.take failed with: out of resources";
+      break;
+    case DDS::RETCODE_NOT_ENABLED:
+      error_string = "_DataReader.take failed with: DataReader is not enabled";
+      break;
+    case DDS::RETCODE_PRECONDITION_NOT_MET:
+      error_string = "_DataReader.take failed with: "
+        "a precondition is not met, one of: "
+        "max_samples > maximum and max_samples != LENGTH_UNLIMITED, or "
+        "the two sequences do not have matching parameters (length, maximum, release), or "
+        "maximum > 0 and release is false.";
+      break;
+    default:
+      error_string = "_DataReader.take failed with unknown return code";
+      break;
+  }
+
+  RMW_SET_ERROR_MSG((std::string("failed to take: ") + prefix + error_string).c_str());
+}
+
+RMW_LOCAL
+rmw_ret_t
+take_serialized_sample(
+  const OpenSpliceStaticSubscriberInfo * subscriber_info,
+  rmw_serialized_message_t * serialized_message,
+  bool ignore_local_publications,
+  bool * taken,
+  DDS::InstanceHandle_t * sending_publication_handle)
+{
+  rmw_ret_t ret = RMW_RET_OK;
+  DDS::CDRSample sample;
+  DDS::SampleInfo info;
+  DDS::DataReader * topic_reader = subscriber_info->topic_reader;
+  DDS::OpenSplice::FooCdrDataReader cdr_reader(topic_reader);
+
+  *taken = false;
+
+  DDS::ReturnCode_t status = cdr_reader.take_cdr(
+    sample, info, DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ANY_INSTANCE_STATE);
+
+  if (status == DDS::RETCODE_OK) {
+    *taken = info.valid_data;
+  } else if (status == DDS::RETCODE_NO_DATA) {
+    *taken = false;
+  } else {
+    report_serialize_take_error(subscriber_info, status);
+    return RMW_RET_ERROR;
+  }
+
+  if (*taken) {
+    DDS::InstanceHandle_t sender_handle = info.publication_handle;
+    auto sender_gid = u_instanceHandleToGID(sender_handle);
+    if (ignore_local_publications) {
+      // compare the system id from the sender and this receiver
+      // if they are equal the sample has been sent from this process and should be ignored
+      DDS::InstanceHandle_t receiver_handle = topic_reader->get_instance_handle();
+      auto receiver_gid = u_instanceHandleToGID(receiver_handle);
+      if (sender_gid.systemId == receiver_gid.systemId) {
+        // ignore sample
+        *taken = false;
+      }
+    }
+    // This is nullptr when being used with plain rmw_take, so check first.
+    if (sending_publication_handle) {
+      *static_cast<DDS::InstanceHandle_t *>(sending_publication_handle) = sender_handle;
+    }
+  }
+
+  if (*taken) {
+    DDS::CDRBlob & blob = sample.blob;
+    auto buffer_size = blob.length();
+    auto ret = rmw_serialized_message_resize(serialized_message, buffer_size);
```

Comment:
I believe it makes sense to resize only when the `serialized_message` instance doesn't have sufficient memory already allocated. It could be that this messages comes from a memory pool which has already pre-allocated memory. In that case, we could safe this call to `resize`.